src/utilities/useEnv.ts

```diff
@@ -0,0 +1,8 @@
+import container from '@/services/container'
```

Comment:
**Question**: The following works correctly with object property lookup:

```ts
import { env } from '@/services/env'

console.log(env.KUMA_DOCS_URL)
```

Why do we need the `useEnv` utility to abstract away this access?

Reply:
Mainly because it looks like most `vue` usage i.e. `useSomething`, I'd probably prefer something like:

```javascript
import { env } from '@/utilities'

console.log(env('KUMA_DOCS_URL'))
```

but I'm trying to stick to Vue's `useSomething()` pattern i.e. as it will probably/maybe help later on

```javascript
import { useEnv } from '@/utilities/env'
const env = useEnv()
console.log(env('KUMA_DOCS_URL'))
```

Reply:
This pattern in Vue has a specific purpose. It’s commonly used for accessing composables which are indeed Vue specific utilities. However, *all* our utility functions are just plain JavaScript. They haven’t got anything to do with Vue. We don’t _need_ to follow a naming pattern of a framework/library we’re using just because it’s otherwise common.

Specifically for composables, we’re talking about utilities that make use of Vue’s state management functionalities (see https://vuejs.org/guide/reusability/composables.html). Composables typically return an object that contains _reactive_ state which is the used _inside a Vue context_. So the composable pattern is used for framework-specific utilities, but it isn’t used for _all_ utilities.

Therefore, **if** we access the `env` utility via a factory function, we shouldn’t use the `use*` naming convention because it’s for composables which `env` isn’t.

---

All that being said, my question is of a different nature: I’m asking why we even need a factory function for doing an object property lookup? We can already directly look up properties on the `env` object. An abstraction around it would be pointless (i.e. it would do that exact same job).

Reply:
I don't think the fact that a 'utility' doesn't use vue specific state management functionalities (at the moment) matters as to whether we happen to call something `use*` or not - more important to me is that everything looks the same (whether the innards use vue specific things right now or not). If we were to change a utility to use vue specific state management features would that mean we should change the name of it to `use*` everywhere, and vice versa?

Honestly, I don't think it's a big deal that this is called `useEnv` or just `getEnv` or whatever, it's just a name. But I'd rather stick to a pattern whether the innards use something special or not. From the outside/as a consumer I don't need to know.

I think another thing is, which I mentioned last time, I've seen other `use*` utilities that don't contain anything vue specific also, so I don't think this is that off the wall.

Reply:
> I don't think the fact that a 'utility' doesn't use vue specific state management functionalities (at the moment) matters as to whether we happen to call something use* or not

It definitely matters to me.

> more important to me is that everything looks the same (whether the innards use vue specific things right now or not).

I don’t think everything should look the same just because that’s technically possible. Using the `use*` naming convention for all utilities would indicate to people familiar with Vue that they’re composables when they’re not.

> If we were to change a utility to use vue specific state management features would that mean we should change the name of it to use* everywhere, and vice versa?

Absolutely yes. Why not? That’s a good distinction to make. Then you know at a glance what’s a pure, plain JavaScript utility function and what relies on Vue internals.

A plain JavaScript function, you can copy as is into any other JavaScript code base and use. You can’t do that with a composable unless that code base also happens to use Vue 3.

These distinctions matter because it gives users of the code (i.e. us) a clear idea of what’s what. Composables and utility functions aren’t on the same hierarchy of things. Composables _are_ utility functions, but utility functions aren’t necessarily composables.

Reply:
> Honestly, I don't think it's a big deal that this is called useEnv or just getEnv or whatever, it's just a name. But I'd rather stick to a pattern whether the innards use something special or not.

I explained [here](https://github.com/kumahq/kuma-gui/pull/507#discussion_r1049649751) that this wasn’t my primary concern.

My primary concern remains that we access the `env` utility via a service abstraction when there is no need for that.

> From the outside/as a consumer I don't need to know.

As a developer, you **have** to know the difference though. You can’t use composables and utilities interchangeably.

> I think another thing is, which I mentioned last time, I've seen other use* utilities that don't contain anything vue specific also, so I don't think this is that off the wall.

That doesn’t align well with the documentation on composables though. They’re described very specifically and narrowly as stateful Vue utilities. https://vuejs.org/guide/reusability/composables.html

Reply:
Oh another thing probably worth mentioning, whilst I'm more than happy to change the name of this here

> Ok so if I change this to getEnv are we good?

Longer term all these accessor functions will hopefully be autogenerated (we're hiding the fact we are using a ServiceContainer), and therefore they'll all use the same naming pattern, if I need to look to see if the function/service is using something Vue specific or not before deciding on a name, we probably won't be able to do the autogeneration.

Just to clarify on the last comment:

> > Ok so if I change this to getEnv are we good?

It sounds like from your last comment, that if I did just change the name here (which I'm happy to do) you still wouldn't be happy with this?

Reply:
> It sounds like from your last comment, that if I did just change the name here (which I'm happy to do) you still wouldn't be happy with this?

I’m afraid so. The functionality in this PR needs a typed object with primitive property values. That covers all the needs of the uses we currently have and that we will have in the short-term future (as far as I’m aware of).

Introducing the proposed service pattern and its abstractions, in its current form, adds logic and code that is not needed and made good use of.

I’m happy to discuss the merits of this pattern as part of a use case that needs it, but I strongly believe accessing static data isn’t one of them.

Reply:
Ok lets put it another way, are you happy that:

1. This uncovered a bunch of tests that should have been failing but weren't (the `null`s).
2. Provides a centralized access point to static environment variables no matter where those actually come from (similar to `useStore()` being for accessing user data, or `useRouter` being for accessing the router)
3. Removes some of the complexity that made it so we were accessing the same value via totally difference approaches in different parts of the codebase.
4. Provides a consistent way to inject alternative data and/or _behaviour_, whether that be within tests or within other different environments.

If we can get past the fact that we are calling a function rather than accessing a property, and naming, we might be able to move forwards a little?

Reply:
> This uncovered a bunch of tests that should have been failing but weren't (the nulls).

Yes, but while that is a small issue, those aren’t representing application bugs, but improperly regenerated snapshots.

> Provides a centralized access point to static environment variables no matter where those actually come from

Yes, I am. I pointed this out in https://github.com/kumahq/kuma-gui/pull/507#pullrequestreview-1219287192

> Removes some of the complexity that made it so we were accessing the same value via totally difference approaches in different parts of the codebase.

Yes, very.

> Provides a consistent way to inject alternative data and/or behaviour, whether that be within tests or within other different environments.

No. I will need better examples for this argument. The current complexity of adding the service container pattern is not warranted in my opinion because the alternative (i.e. accessing property values from an object) is strictly simpler, requires less code, and has no disadvantages. I’m not bringing this up because of stylistic preference of how this data is accessed. I’m opposed to adding abstractions to things that don’t need them and accessing static data like we need to really doesn’t.

I’d like to ask you to propose this pattern in context of a justified use case and outside the scope of this PR.

Reply:
> No. I will need better examples for this argument.

Here's one example, we are injecting functionality, not static values, and not using our current (but soon to change) test utilities, i.e. we could just as well do this for a different environment where we don't have test utilities available.

https://github.com/kumahq/kuma-gui/pull/507/files#diff-5cb6e61dca3b13ba5c02a1b586af105e5e92eb5d080e3cea1a4ccc8bf86cd738R14

> The current complexity of adding the service container pattern is not warranted in my opinion

We use a similar approach on the backend to avoid merge conflicts, the GUI currently has a merge conflict problem, so this approach is a common solution to that, I would say it's warranted (either this or another alternative). I think the bigger picture is, we need to inject behaviour aswell as just static variables, unless we are happy with conditonal-y feature flags throughout the codebase, something I thought we were in agreement that we had to avoid.

I get that this could be done simpler, by just having an object and hanging some properties off of it, but as I said up there ^ - this is about more than that. And even then what we have here isn't really that complex, and I've talked about some of the advantages of the overall approach. I'm also a little confused as you were happy to approve this https://github.com/kumahq/kuma-gui/pull/466#pullrequestreview-1208150922 with reservations (but different ones) and we've briefly chatted about this approach offline. Overall if you want to put more discussion into this, I'd be happy to but its probably going to have to wait now 😿 

If you'd rather leave this hanging for today, please let me know and we can continue the discussion another time. I'm easy either way.